core/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala

```diff
@@ -213,28 +232,104 @@ case class UpdateCommand(
       spark, txn, "update", rootPath, inputLeafFiles, nameToAddFileMap)
     val newTarget = DeltaTableUtils.replaceFileIndex(target, baseRelation.location)
     val targetDf = Dataset.ofRows(spark, newTarget)
-    val updatedDataFrame = {
-      val updatedColumns = buildUpdatedColumns(condition)
-      targetDf.select(updatedColumns: _*)
-    }
+
+    val updatedDataFrame = withUpdatedColumns(
+      target,
+      updateExpressions,
+      condition,
+      targetDf.withColumn(CONDITION_COLUMN_NAME, new Column(condition)),
+      shouldOutputCdc(txn))
 
     txn.writeFiles(updatedDataFrame)
   }
+}
+
+object UpdateCommand {
+  val FILE_NAME_COLUMN = "_input_file_name_"
+  val CONDITION_COLUMN_NAME = "__condition__"
+
+  /**
+   * Whether or not CDC is enabled on this table and, thus, if we should output CDC data during this
+   * UPDATE operation.
+   */
+  def shouldOutputCdc(txn: OptimisticTransaction): Boolean = {
+    DeltaConfigs.CHANGE_DATA_FEED.fromMetaData(txn.metadata)
+  }
 
   /**
    * Build the new columns. If the condition matches, generate the new value using
-   * the corresponding UPDATE EXPRESSION; otherwise, keep the original column value
+   * the corresponding UPDATE EXPRESSION; otherwise, keep the original column value.
+   *
+   * When CDC is enabled, includes the generation of CDC preimage and postimage columns for
+   * changed rows.
+   *
+   * @param target target we are updating into
+   * @param updateExpressions the update transformation to perform on the input DataFrame
+   * @param metrics transaction metrics
+   * @param dfWithEvaluatedCondition source DataFrame on which we will apply the update condition
```

Comment:
isn't this DataFrame already has `CONDITION_COLUMN_NAME`? Why does the argument description say "on which we will apply the update condition with an additional column"?

Reply:
i'll change it to "on which we will apply the update expression". the evaluated condition is there (true or false), but the actual update on the row has not been performed.